packages/sources/tiingo/src/adapter.ts

```diff
@@ -36,3 +42,83 @@ export const execute: ExecuteWithConfig<Config> = async (request, config) => {
 export const makeExecute: ExecuteFactory<Config> = (config) => {
   return async (request) => execute(request, config || makeConfig())
 }
+
+type MessageType =
+  | 'A' // for new data
+  | 'U' // for updating existing data
+  | 'D' // for deleing existing data
+  | 'I' // for informational/meta data
+  | 'E' // for error messages
+  | 'H' // for Heartbeats (can be ignored for most cases)
+interface Message {
+  service: string
+  response: { message: string; code: number }
+  messageType: MessageType
+}
+interface TopOfBookUpdateData {
+  0: 'Q'
+  1: string // Ticker related to the asset.
+  2: string // A string representing the datetime this trade quote came in.
+  3: string // The exchange the top-of-book update is for.
+  4: number // The amount of crypto at the bid price in the base currency.
+  5: number // The current highest bid price.
+  6: number // The mid price of the current timestamp when both "bidPrice" and "askPrice" are not-null. In mathematical terms: mid = (bidPrice + askPrice)/2.0
+  7: number // The amount of crypto at the ask price in the base currency.
+  8: number // The current lowest ask price.
+}
+interface TradeUpdateData {
+  0: 'T'
+  1: string // Ticker related to the asset.
+  2: string // A string representing the datetime this trade quote came in.
+  3: string // The exchange the trade was done one.
+  4: number // The amount of crypto volume done at the last price in the base currency.
+  5: number // The last price the last trade was executed at.
+}
+interface UpdateMessage extends Message {
+  messageType: 'A'
+  service: 'crypto_data'
+  data: TopOfBookUpdateData | TradeUpdateData
+}
+
+const customParams = {
+  base: ['base', 'from', 'coin'],
+  quote: ['quote', 'to', 'market'],
+}
+
+export const makeWSHandler = (config?: Config): MakeWSHandler => {
+  const defaultConfig = config || makeConfig()
+  const getSubscription = (pair: string | undefined, subscribe = true) => {
+    if (!pair) return
+    return {
+      eventName: subscribe ? 'subscribe' : 'unsubscribe',
+      authorization: defaultConfig?.apiKey,
+      eventData: {
+        thresholdLevel: 5, // only Last Trade updates
+        tickers: [pair],
+      },
+    }
+  }
+  const getPair = (input: AdapterRequest) => {
+    const validator = new Validator(input, customParams)
+    if (validator.error) return
+    const base = validator.validated.data.base.toLowerCase()
+    const quote = validator.validated.data.quote.toLowerCase()
+    return `${base}${quote}`
+  }
+  return () => {
```

Comment:
Include the `const defaultConfig = config || makeConfig()` here better. This is useful when other adapters rely on this one

Reply:
Not sure that I understand how this would be any different than using it as a closure as `makeConfig` doesn't change, but I moved it so we can get this merged.